src/net_processing.cpp

```diff
@@ -5817,6 +5817,12 @@ bool PeerManagerImpl::SendMessages(CNode* pto)
     // disconnect misbehaving peers even before the version handshake is complete.
     if (MaybeDiscourageAndDisconnect(*pto, *peer)) return true;
 
+    // Initiate version handshake for outbound connections
+    if (!pto->IsInboundConn() && !peer->m_initial_version_message_sent) {
+        PushNodeVersion(*pto, *peer);
+        peer->m_initial_version_message_sent = true;
+    }
```

Comment:
This isn't fully atomic, i.e. in theory two threads could enter this section at the same time.

Not tested but maybe the following works:
```suggestion
    if (!pto->IsInboundConn() && peer->m_initial_version_message_sent.compare_exchange_strong(false, true)) {
        PushNodeVersion(*pto, *peer);
    }
```

Alternatively, we could guard `m_initial_version_message_sent` with `NetEventsInterface::g_msgproc_mutex`, since it is only accessed from the "msghand" thread.